File: frontend/src/types/health.ts
// Health API types (DEC-042)

export type SubsystemStatus = 'healthy' | 'degraded' | 'unhealthy'
export type OverallStatus = 'healthy' | 'degraded' | 'unhealthy'

export interface DatabaseStatus {
  status: SubsystemStatus
  latency_ms: number | null
}

export interface TelegramStatus {
  status: SubsystemStatus
  connected: boolean
  authenticated: boolean
}

export interface WorkersStatus {
  status: SubsystemStatus
  jobs_queued: number
  jobs_running: number
  jobs_failed_24h: number
}

export interface StorageStatus {
  status: SubsystemStatus
  library_gb: number
  staging_gb: number
  cache_gb: number
  library_free_gb: number
}

export interface RateLimiterStatus {
  status: SubsystemStatus
  requests_total: number
  throttled_count: number
  channels_in_backoff: number
}

export interface Subsystems {
  database: DatabaseStatus
  telegram: TelegramStatus
  workers: WorkersStatus
  storage: StorageStatus
  rate_limiter: RateLimiterStatus
}

export interface RecentError {
  job_id: string
  job_type: string
  error: string
  timestamp: string
}

export interface DetailedHealthResponse {
  overall: OverallStatus
  version: string
  subsystems: Subsystems
  errors: RecentError[]
}

File: backend/app/schemas/health.py
"""Health check schemas."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel


class HealthResponse(BaseModel):
    """Health check response schema."""

    status: str
    version: str
    database: str  # "connected" or "disconnected"


# ==================== Detailed Health Check Schemas (DEC-042) ====================


class SubsystemStatus(BaseModel):
    """Base status for a subsystem."""

    status: Literal["healthy", "degraded", "unhealthy"]


class DatabaseStatus(SubsystemStatus):
    """Database subsystem status."""

    latency_ms: float | None = None


class TelegramStatus(SubsystemStatus):
    """Telegram subsystem status."""

    connected: bool = False
    authenticated: bool = False


class WorkersStatus(SubsystemStatus):
    """Workers subsystem status."""

    jobs_queued: int = 0
    jobs_running: int = 0
    jobs_failed_24h: int = 0


class StorageStatus(SubsystemStatus):
    """Storage subsystem status."""

    library_gb: float = 0.0
    staging_gb: float = 0.0
    cache_gb: float = 0.0
    library_free_gb: float = 0.0


class RateLimiterStatus(SubsystemStatus):
    """Rate limiter subsystem status."""

    requests_total: int = 0
    throttled_count: int = 0
    channels_in_backoff: int = 0


class Subsystems(BaseModel):
    """All subsystem statuses."""

    database: DatabaseStatus
    telegram: TelegramStatus
    workers: WorkersStatus
    storage: StorageStatus
    rate_limiter: RateLimiterStatus


class RecentError(BaseModel):
    """Recent error entry."""

    job_id: str
    job_type: str
    error: str
    timestamp: str


class DetailedHealthResponse(BaseModel):
    """Detailed health check response (DEC-042)."""

    overall: Literal["healthy", "degraded", "unhealthy"]
    version: str
    subsystems: Subsystems
    errors: list[RecentError] = []
